src/nvim/api/autocmd.c

```diff
@@ -0,0 +1,669 @@
+#include <stdbool.h>
+#include <stdio.h>
+
+#include "lauxlib.h"
+#include "nvim/api/autocmd.h"
+#include "nvim/api/private/defs.h"
+#include "nvim/api/private/helpers.h"
+#include "nvim/ascii.h"
+#include "nvim/buffer.h"
+#include "nvim/eval/typval.h"
+#include "nvim/fileio.h"
+#include "nvim/lua/executor.h"
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "api/autocmd.c.generated.h"
+#endif
+
+#define AUCMD_MAX_PATTERNS 256
+
+// Check whether every item in the array is a kObjectTypeString
+#define CHECK_STRING_ARRAY(__array, k, v, goto_name) \
+  for (size_t j = 0; j < __array.size; j++) { \
+    Object item = __array.items[j]; \
+    if (item.type != kObjectTypeString) { \
+      api_set_error(err, \
+                    kErrorTypeValidation, \
+                    "All entries in '%s' must be strings", \
+                    k); \
+      goto goto_name; \
+    } \
+  }
+
+// Copy string or array of strings into an empty array.
+#define UNPACK_STRING_OR_ARRAY(__array, k, v, goto_name) \
```

Comment:
this assumes the array has not been set before? So it could be a function returning the array and check `ERROR_SET(err)` for error in caller :]

Reply:
It's also only using `goto cleanup` so far, so it can also be converted to an inline function.